Assets/Mirror/Editor/Weaver/CompilationFinishedHook.cs

```diff
@@ -0,0 +1,147 @@
+        {
+            const string k_HlapiRuntimeAssemblyName = "Mirror";
+
+            // Do nothing if there were compile errors on the target
+            if (CompilerMessagesContainError(messages))
+            {
+                Debug.Log("Weaver: stop because compile errors on target");
+                return;
+            }
+
+            // Should not run on the editor only assemblies
+            if (assemblyPath.Contains("-Editor") || assemblyPath.Contains(".Editor"))
+            {
+                return;
+            }
+
+            // don't weave mirror files
+            string assemblyName = Path.GetFileName(assemblyPath);
+            if (assemblyName == "Telepathy.dll" || assemblyName == "Mirror.dll" || assemblyName == "Mirror.Weaver.dll")
+            {
+                return;
+            }
+
+            string mirrorRuntimeDll = FindMirrorRuntime();
+            if (string.IsNullOrEmpty(mirrorRuntimeDll))
+            {
+                Debug.LogError("Failed to find Mirror runtime assembly");
+                return;
+            }
+            if (!File.Exists(mirrorRuntimeDll))
+            {
+                // this is normal, it happens with any assembly that is built before mirror
+                // such as unity packages or your own assemblies
+                // those don't need to be weaved
+                // if any assembly depends on mirror, then it will be built after
+                return;
+            }
+
+            Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies();
+
+            string unityEngineCoreModuleDLL = UnityEditorInternal.InternalEditorUtility.GetEngineCoreModuleAssemblyPath();
+            if (string.IsNullOrEmpty(unityEngineCoreModuleDLL))
+            {
+                Debug.LogError("Failed to find UnityEngine assembly");
+                return;
+            }
+
+            bool usesMirror = false;
+            bool foundThisAssembly = assemblies.Any(assembly => assembly.GetName().Name == Path.GetFileNameWithoutExtension(assemblyPath));
+            HashSet<string> dependencyPaths = new HashSet<string>();
+            foreach (Assembly assembly in assemblies)
+            {
+                // Find the assembly currently being compiled from domain assembly list and check if it's using Mirror
+                if (assembly.GetName().Name == Path.GetFileNameWithoutExtension(assemblyPath))
+                {
+                    foreach (AssemblyName dependency in assembly.GetReferencedAssemblies())
+                    {
+                        // Since this assembly is already loaded in the domain this is a no-op and returns the
+                        // already loaded assembly
+                        string location = Assembly.Load(dependency).Location;
+                        dependencyPaths.Add(Path.GetDirectoryName(location));
+                        if (dependency.Name.Contains(k_HlapiRuntimeAssemblyName))
```

Comment:
this is a little to broad,   you are saying if any dependency contains the word "Mirror",  then it uses mirror.    You should check if the dependency is Mirror.dll specifically.  Someone someday will have an asset with the word Mirror on it, and we will consider it here.

Reply:
that's from unity's code. still refactoring it.